src/flag_gems/utils/shape_utils.py

```diff
@@ -212,37 +212,48 @@ def add_on_kernel(
 ):
     pid_x = tl.program_id(axis=0)
     pid_y = tl.program_id(axis=1)
-    rows_offset = pid_x * BLOCK_M + tl.arange(0, BLOCK_M)[:, None]
+    rows_offset = pid_x * BLOCK_M + tl.arange(0, BLOCK_M)
     rows_mask = rows_offset < M
 
-    cols_offset = pid_y + tl.arange(0, BLOCK_N)[None, :]
+    cols_offset = pid_y + tl.arange(0, BLOCK_N)
     cols_mask = cols_offset < N
-    block_mask = rows_mask and cols_mask
+    block_mask = rows_mask[:, None] & cols_mask[None, :]
+
+    offsets = rows_offset[:, None] * N + cols_offset[None, :]
+    cur_idx = tl.load(idx + offsets, mask=block_mask, other=0)
 
-    offsets = rows_offset * N + cols_offset
-    cur_idx = tl.load(idx + offsets, mask=block_mask, other=1)
     mod = cur_idx % cur_shape
     res = mod * cur_strides
+
     tl.store(add_on + offsets, res, mask=block_mask)
 
 
 def offset_calculator(inp, idx, strides, dim, isInp):
     ndim = inp.ndim
     shape = list(inp.shape)
+    idx = idx.contiguous()
+
     offsets = torch.zeros_like(inp, dtype=torch.int32, device=inp.device)
     idx_dim = torch.zeros_like(inp, dtype=torch.int32, device=inp.device)
-    for d in range(0, ndim):
+
+    N = idx.size(idx.ndim - 1)
+    M = idx.numel() // N
+
+    # Ensure grid configuration is correct
+    grid = lambda meta: (
+        triton.cdiv(M, meta["BLOCK_M"]),
+        triton.cdiv(N, meta["BLOCK_N"]),
+    )
+
+    for d in range(ndim):
+        # Ensure correct size for add_on
         add_on = torch.zeros_like(inp, dtype=torch.int32, device=inp.device)
```

Comment:
Since idx is in int64, int32 add_on may trigger overflow. could it be the cause?

Reply:
I tend to say not because I've already considered this reason and made the corresponding modifications before, without fixing the original issue.